go/vt/vtgate/safe_session.go

```diff
@@ -32,11 +32,36 @@ import (
 // (the use pattern is 'Find', if not found, then 'Append',
 // for a single shard)
 type SafeSession struct {
-	mu           sync.Mutex
-	mustRollback bool
+	mu              sync.Mutex
+	mustRollback    bool
+	autocommitState autocommitState
 	*vtgatepb.Session
 }
 
+// autocommitState keeps track of whether a single round-trip
+// commit to vttablet is possible. It starts as autocommitable
+// if we started a transaction because of the autocommit flag
+// being set. Otherwise, it starts as notAutocommitable.
+// If execute is recursively called using the same session,
+// like from a vindex, we will already be in a transaction,
+// and this should cause the state to become notAutocommitable.
+//
+// SafeSession lets you request a commit token, which will
+// be issued if the state is autocommitable, and ShardSessions
+// is empty, implying that no intermediate transactions were started.
+// If so, the state transitions to autocommited, which is terminal.
+// If the token is succesfully issued, the caller has to perform
+// the commit. If a token cannot be issued, then a traditional
+// commit has to be performed at the outermost level where
+// the autocommitable transition happened.
+type autocommitState int
+
+const (
+	notAutocommitable = autocommitState(iota)
```

Comment:
nit -- these should all be `notAutocommittable`, `autocommittable`, and `autocommitted` to match the actual english spelling of "committed" / "committable".